fem/tmop.cpp

```diff
@@ -0,0 +1,2283 @@
+   dM *= 1.0 / pow(det, 4./3.);
+}
+
+// dI2_dM = [ -4/3 |adj(M)|^2  det(M)^(1/3) adj(M)^T ] / det(M)^(8/3).
+void TMOP_QualityMetric::Dim3Invariant2_dM(const DenseMatrix &M,
+                                           DenseMatrix &dM)
+{
+   MFEM_ASSERT(M.Height() == 3 && M.Width() == 3, "Incorrect dimensions!");
+
+   DenseMatrix Madj(3);
+   // dM will have Madj^t because it is the third invariant's derivative.
+   CalcAdjugate(M, Madj);
+   const double fnorm = Madj.FNorm(), det = M.Det();
+
+   Dim3Invariant3_dM(M, dM);
+   dM *= -(4./3.)* fnorm * fnorm * pow(det, 1./3.);
+   dM *= 1.0 / (pow(det, 8./3.));
+}
+
+// dI3_dM = d(det(M))_dM = adj(M)^T.
+void TMOP_QualityMetric::Dim3Invariant3_dM(const DenseMatrix &M,
+                                           DenseMatrix &dM)
+{
+   MFEM_ASSERT(M.Height() == 3 && M.Width() == 3, "Incorrect dimensions!");
+
+   CalcAdjugateTranspose(M, dM);
+}
+
+void TMOP_QualityMetric::Dim3Invariant1_dMdM(const DenseMatrix &M, int i, int j,
+                                             DenseMatrix &dMdM)
+{
+   MFEM_ASSERT(M.Height() == 3 && M.Width() == 3, "Incorrect dimensions!");
+
+   DenseMatrix dI(3);
+   Dim3Invariant3_dM(M, dI);
+   const double fnorm  = M.FNorm(), det = M.Det();
+
+   DenseMatrix dM(3); dM = 0.0; dM(i, j) = 1.0;
+   for (int r = 0; r < 3; r++)
+   {
+      for (int c = 0; c < 3; c++)
+      {
+         dMdM(r,c) = (2.0 * det * det * dM(r,c)
+                      + dI(i,j) * (10./9.) * fnorm * fnorm * dI(r,c)
+                      - (4./3.) * dI(i,j) * det * M(r,c)
+                      - (4./3.) * det * M(i,j) * dI(r,c))
+                     / pow(det, 8./3.);
+      }
+   }
+}
+
+void TMOP_QualityMetric::Dim3Invariant2_dMdM(const DenseMatrix &M, int i, int j,
+                                             DenseMatrix &dMdM)
+{
+   MFEM_ASSERT(M.Height() == 3 && M.Width() == 3, "Incorrect dimensions!");
+
+   DenseMatrix dI(3);
+   Dim3Invariant3_dM(M, dI);
+   DenseMatrix Madj(3);
+   CalcAdjugate(M, Madj);
+   const double det   = M.Det();
+   const double fnorm = Madj.FNorm();
+
+   DenseMatrix dM(3); dM = 0.0; dM(i, j) = 1.0;
+   for (int r = 0; r < 3; r++)
+   {
+      for (int c = 0; c < 3; c++)
+      {
+         dMdM(r,c) = (28./9.) * fnorm * fnorm *
+                     det * det * dI(i,j) * dI(r,c) / pow(det, 16./3.);
+      }
+   }
+}
+
+// (dI3_dM)_d(Mij) = 0.
+void TMOP_QualityMetric::Dim3Invariant3_dMdM(const DenseMatrix &M, int i, int j,
+                                             DenseMatrix &dMdM)
+{
+   MFEM_ASSERT(M.Height() == 3 && M.Width() == 3, "Incorrect dimensions!");
+
+   dMdM(i, j) = 0.0;
+}
+
+
+double TMOP_Metric_001::EvalW(const DenseMatrix &Jpt) const
+{
+#ifdef MFEM_USE_OLD_TMOP_INVARIANTS
+   return Dim2Invariant1(Jpt) * Dim2Invariant2(Jpt);
+#else
+   ie.SetJacobian(Jpt.GetData());
+   return ie.Get_I1();
+#endif
+}
+
+void TMOP_Metric_001::EvalP(const DenseMatrix &Jpt, DenseMatrix &P) const
+{
+#ifdef MFEM_USE_OLD_TMOP_INVARIANTS
```

Comment:
Consider removing these?

Reply:
Yes, I'll remove the old implementation.